tools/scylla-nodetool.cc

```diff
@@ -2091,6 +2091,16 @@ void status_operation(scylla_rest_client& client, const bpo::variables_map& vm)
         dc_endpoints[dc].insert(ep);
     }
 
+    for (const auto& [ep, host_id] : endpoint_host_id) {
+        if (endpoint_rack.contains(ep)) {
+            continue;
+        }
+        const auto dc = sstring(rjson::to_string_view(client.get("/snitch/datacenter", {{"host", ep}})));
+        const auto rack = sstring(rjson::to_string_view(client.get("/snitch/rack", {{"host", ep}})));
+        endpoint_rack.emplace(ep, rack);
+        dc_endpoints[dc].insert(ep);
+    }
+
```

Comment:
> IIUC, all nodes appear in the new loop,

For this to be true, we need to always have host_id associated with nodes, which is sometimes not true for joining nodes. We have a test in test_status where host_id is unavailable. But this isn't a problem for zero token nodes since we anyway dont display joining nodes there. 

Hence a union of `tokens_endpoint` and `host_id_map` serves our purpose in all cases.

```
def test_status_keyspace_joining_node(request, nodetool):
    """ Joining nodes do not have some attributes available yet:
    * load
    * host_id
    """
```

Reply:
> For this to be true, we need to always have host_id associated with nodes, which is sometimes not true for joining nodes. 

Which is a bug: https://github.com/scylladb/scylladb/issues/17857
I recall that we discussed it and I asked you to fix it as part of this PR, Abhinav.

> We have a test in test_status where host_id is unavailable.

So we're verifying that a bug is present? That doesn't make sense.

Reply:
The `/storage_service/host_id` api is calling `token_metadata::get_endpoint_to_host_id_map_for_reading` underneath.

This is how the function is implemented:
```cpp
std::unordered_map<inet_address, host_id> token_metadata_impl::get_endpoint_to_host_id_map_for_reading() const {
    const auto& nodes = _topology.get_nodes_by_endpoint();
    std::unordered_map<inet_address, host_id> map;
    map.reserve(nodes.size());
    for (const auto& [endpoint, node] : nodes) {
        // Restrict to members
        if (!node->is_member()) {
            continue;
        }
        if (const auto& host_id = node->host_id()) {
            map[endpoint] = host_id;
        } else {
            tlogger.info("get_endpoint_to_host_id_map_for_reading: endpoint {} has null host_id: state={}", endpoint, node->get_state());
        }
    }
    return map;
}
```

As we can see, it is skipping joining nodes:
```cpp
        // Restrict to members
        if (!node->is_member()) {
            continue;
        }
```

And perhaps that made sense at some point. The function name suffix `_for_reading` suggests that the function was used in some other places in the past, and indeed if we need endpoints "for reading" then we cannot show joining endpoints.

However, recall @AbhinavKumarJha that  we determined that this function is used **only** by this api endpoint today. So we can modify it as we like to suit our needs, and also change the name to better describe what it does after our fixes.


So if we modify the endpoint to include joining nodes, then joining nodes will appear in the `/storage_service/host_id` endpoint. Then, as @patjed41 observed, you should not need the first loop: all nodes will appear in the `for (const auto& [ep, host_id] : endpoint_host_id) {` loop. 

At the same time, this will fix https://github.com/scylladb/scylladb/issues/17857. Two birds with one stone!

Now my question is: in earlier version of this PR, I saw that you actually modified that `get_endpoint_to_host_id_map_for_reading` function. So it seems that you actually tried my proposal. So perhaps I'm missing something and it doesn't work?

Reply:
>So we're verifying that a bug is present? That doesn't make sense

The test verifies that even if host_id is None, the nodetool will display the node. Since we already discussed that this case can never exist, I added `@pytest.mark.skip(reason="This scenario can never happen, all joining nodes deterministically have host_id available")` for that test.(Since we use host_id_map, this test fails with our implementation)

> At the same time, this will fix https://github.com/scylladb/scylladb/issues/17857. Two birds with one stone!

Sure, done

Reply:
> The test verifies that even if host_id is None, the nodetool will display the node. Since we already discussed that this case can never exist, I added @pytest.mark.skip(reason="This scenario can never happen, all joining nodes deterministically have host_id available") for that test

We could keep it enabled, you could e.g. run newer nodetool against older version of Scylla, which still has https://github.com/scylladb/scylladb/issues/17857, and then the test verifies that in this case, nodetool will still show nodes with missing host IDs.

> Since we use host_id_map, this test fails with our implementation

It fails? Then I don't understand something:
- does the test run against mocked REST API endpoints? If it does, then it shouldn't matter what we do inside Scylla, the test should still pass because the mocked endpoint does not return host ID
- or does the test run against a real Scylla instance?

Reply:
As already discussed, I removed the test. Reasoning:

Tests are run against mock API endpoints only. For that particular test, host_id_map returned from API has not changed. What has changed is that earlier we used token_endpoint_map for discovering what nodes are present(lets call it v1). Now we are using endpoint_host_id map for that purpose(lets call it v2).
Now, the test that I am talking about has host_id=None which means that in v1, it would show the host_id as ?, but for v2 it won't display that node at all, since we are using endpoint_host_id map.